silhouette/Graphtec.py

```diff
@@ -458,331 +619,355 @@ def wait_for_ready(s, timeout=30, verbose=True):
-      #TB50,x: x = 1 landscape mode, x = 0 portrait mode
-      s.write(b"FN0\x03TB50,0\x03")
+    if self.dev is None: return None
 
-      if cuttingmat == 'cameo_12x12':
-        s.write(b"\\%d,%d\x03Z%d,%d\x03" % (0, 0, 6096, 6096))
-      elif cuttingmat == 'cameo_12x24':
-        s.write(b"\\%d,%d\x03Z%d,%d\x03" % (0, 0, 12192, 6096))
-      else:
-        width = s.hardware['width_mm'] if 'width_mm' in s.hardware else mediawidth
-        height = s.hardware['length_mm'] if 'length_mm' in s.hardware else mediaheight
-        s.write(b"\\%d,%d\x03Z%d,%d\x03" % (0, 0, height * 20.0, width * 20.0))
+    self.initialize()
+
+    self.set_cutting_mat(cuttingmat, mediawidth, mediaheight)
 
     if media is not None:
       if media < 100 or media > 300: media = 300
-      if s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO3 or s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO4:
-        # Silhouette Studio does not appear to issue this command
-        pass
-      else:
-        s.write(b"FW%d\x03" % media);
+
+      # Silhouette Studio does not appear to issue this command
+      if self.product_id() not in [PRODUCT_ID_SILHOUETTE_CAMEO3, PRODUCT_ID_SILHOUETTE_CAMEO4]:
+        self.send_command("FW%d" % media)
+
       if pen is None:
         if media == 113:
           pen = True
         else:
           pen = False
       for i in MEDIA:
         if i[0] == media:
-          print("Media=%d, cap='%s', name='%s'" % (media, i[4], i[5]), file=s.log)
+          print("Media=%d, cap='%s', name='%s'" % (media, i[4], i[5]), file=self.log)
           if pressure is None: pressure = i[1]
           if    speed is None:    speed = i[2]
           if    depth is None:    depth = i[3]
           break
 
+    tool = SilhouetteCameoTool(toolholder)
+
     if toolholder is None:
       toolholder = 1
-    if s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO3 or s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO4:
-      s.write(b"J%d\x03" % toolholder)
-    print("toolholder: %d" % toolholder, file=s.log)
+
+    if self.product_id() in [PRODUCT_ID_SILHOUETTE_CAMEO3, PRODUCT_ID_SILHOUETTE_CAMEO4]:
+      self.send_command(tool.select())
+
+    print("toolholder: %d" % toolholder, file=self.log)
 
     # cameo 4 sets some parameters two times (force, TJ%d, Cutter offset)
-    if s.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO4:
+    if self.product_id() == PRODUCT_ID_SILHOUETTE_CAMEO4:
       if pressure is not None:
         if pressure <  1: pressure = 1
         if pressure > 33: pressure = 33
-        s.write(b"FX%d,%d\x03" % (pressure, toolholder));
-        print("pressure: %d" % pressure, file=s.log)
+        self.send_command(tool.pressure(pressure))
+        print("pressure: %d" % pressure, file=self.log)
+
         # some sort of pressure calibration I presume (only 0 on first connection to cameo else 3)
-        # hopefully also allowed later on
-        s.write(b"TJ0\x03")
+        # hopefully also allowed later on, maybe accelleration
```

Comment:
You are right off course TJ transmits the acceleration for Cameo4. But as for the FF_,_,_ command it first sends the previously used value (0 after first connection) and then the new value used for the upcoming plot. So I guess as with the FF we should keep the double sending of the command and also as with FF always send 0 with the first command as if it was reconnected.